applications/GeoMechanicsApplication/custom_strategies/schemes/geomechanics_time_integration_scheme.hpp

```diff
@@ -302,19 +344,83 @@ class GeoMechanicsTimeIntegrationScheme : public Scheme<TSparseSpace, TDenseSpac
     {
         KRATOS_TRY
 
-        block_for_each(rDofSet, [&Dx](auto& dof) {
-            if (dof.IsFree()) {
+        block_for_each(rDofSet, [&Dx](auto& dof)
+        {
+            if (dof.IsFree())
+            {
                 dof.GetSolutionStepValue() +=
                     TSparseSpace::GetValue(Dx, dof.EquationId());
             }
         });
 
-        this->UpdateVariablesDerivatives(rModelPart);
+        UpdateVariablesDerivatives(rModelPart);
 
         KRATOS_CATCH("")
     }
 
 protected:
+    const Variable<double>& GetComponentFromVectorVariable(
+        const Variable<array_1d<double, 3>>& rSource, const std::string& rComponent) const
+    {
+        return KratosComponents<Variable<double>>::Get(rSource.Name() + "_" + rComponent);
+    }
+
+    virtual inline void SetTimeFactors(ModelPart& rModelPart)
+    {
+        mDeltaTime = rModelPart.GetProcessInfo()[DELTA_TIME];
+    }
+
+    virtual inline void UpdateVariablesDerivatives(ModelPart& rModelPart) = 0;
+
+    [[nodiscard]] double GetDeltaTime() const { return mDeltaTime; }
+
+    [[nodiscard]] const std::vector<SecondOrderVectorVariable>& GetSecondOrderVectorVariables() const
+    {
+        return mSecondOrderVectorVariables;
+    }
+
+    [[nodiscard]] const std::vector<FirstOrderScalarVariable>& GetFirstOrderScalarVariables() const
+    {
+        return mFirstOrderScalarVariables;
+    }
+
+private:
+    void CheckAllocatedVariables(const ModelPart& rModelPart) const
+    {
+        for (const auto& r_node : rModelPart.Nodes())
+        {
+            for (const auto& r_first_order_scalar_variable : mFirstOrderScalarVariables)
+            {
+                this->CheckSolutionStepsData(r_node, r_first_order_scalar_variable.instance);
+                this->CheckSolutionStepsData(r_node, r_first_order_scalar_variable.first_time_derivative);
+                this->CheckDof(r_node, r_first_order_scalar_variable.instance);
+            }
+        }
+
+        for (const auto& r_node : rModelPart.Nodes())
+        {
+            for (const auto& r_second_order_vector_variable : this->mSecondOrderVectorVariables)
+            {
+                if (!rModelPart.HasNodalSolutionStepVariable(
+                        r_second_order_vector_variable.instance))
+                    continue;
+
+                this->CheckSolutionStepsData(r_node, r_second_order_vector_variable.instance);
+                this->CheckSolutionStepsData(r_node, r_second_order_vector_variable.first_time_derivative);
+                this->CheckSolutionStepsData(r_node, r_second_order_vector_variable.second_time_derivative);
+
+                // We don't check for "Z", since it is optional (in case of a 2D problem)
```

Comment:
This might still be a point of discussion

Reply:
As discussed, let's do that at a different moment, since we have to find out where to get the information from if the case is 2D/3D